toolchain/install/install_paths.h

```diff
@@ -0,0 +1,112 @@
+// support locating an install through Bazel's runfiles tree or through an
+// explicit path for other use cases. When locating an install, we verify it by
+// looking for the `carbon_install.txt` marker file at a specific location
+// below. The install paths object retains any error information so that the
+// driver can diagnose errors and report them as needed, but continue to
+// function minimally. No methods will crash even in an error state, they will
+// just return based on an empty install prefix.
+//
+// Within this prefix, we expect a hierarchy on Unix-y platforms:
+//
+// - `prefix_root/bin/carbon` - the main CLI driver
+// - `prefix_root/lib/carbon/carbon_install.txt` - a marker for the install
+// - `prefix_root/lib/carbon/...` - private data & binaries
+//
+// This is loosely based on the FHS (Filesystem Hierarchy Standard).
+//
+// An instance of this class provides methods that query for specific paths
+// within the install. Note that we want to abstract away any platform
+// differences in the installation layout, and so while there are some broad
+// paths available here, like the `prefix` method, those should primarily be
+// used for logging or debugging. When a specific part of the install is needed,
+// a dedicated accessor should be added that computes the path for that
+// component.
+//
+// Path accessor methods on the class return `llvm::StringRef` for any paths
+// that are stored in the class, and a `std::string` for any that are computed
+// on demand.
+//
+// TODO: Need to check the installation structure of LLVM on Windows and figure
+// out what Carbon's should be within a Windows prefix and how much of the
+// structure we can share with the Unix-y layout of the prefix.
+class InstallPaths {
+ public:
+  // Provide the current executable's path to detect the correct installation
+  // prefix path. This requires the toolchain to be in its installed layout.
+  static auto MakeExeRelative(llvm::StringRef exe_path) -> InstallPaths;
+
+  // Provide the current executable's path, and use that to detect a Bazel or
+  // Bazel-compatible runfiles install prefix path. This should only be used
+  // where it is reasonable to rely on this rather than a fixed install location
+  // such as for internal development purposes or other Bazel users of the
+  // Carbon library.
+  static auto MakeForBazelRunfiles(llvm::StringRef exe_path) -> InstallPaths;
+
+  // Provide an explicit install paths prefix. This is useful for testing or for
+  // using Carbon in an environment with an unusual path to the installed files.
+  static auto Make(llvm::StringRef install_prefix) -> InstallPaths;
+
+  // Check for an error detecting the install paths correctly.
+  //
+  // An empty return means no errors encountered and the paths should work
+  // correctly.
+  //
+  // A non-empty return means there was an error, and details of the error are
+  // in the `StringRef` for inclusion in any user report.
```

Comment:
```suggestion
  // Check for an error detecting the install paths correctly.
  //
  // A nullopt return means no errors encountered and the paths should work
  // correctly.
  //
  // A string return means there was an error, and details of the error are
  // in the `StringRef` for inclusion in any user report.
```

Was this originally written as returning a StringRef directly, not optional?

Reply:
Doh, yes. Thanks. =]